onnxruntime/core/session/inference_session.cc

```diff
@@ -97,37 +98,172 @@ inline std::basic_string<T> GetCurrentTimeString() {
-      thread_pool_(concurrency::CreateThreadPool("intra_op_thread_pool",
-                                                 session_options.intra_op_num_threads)),
-      inter_op_thread_pool_(session_options.execution_mode == ExecutionMode::ORT_PARALLEL
-                                ? concurrency::CreateThreadPool("inter_op_thread_pool",
-                                                                session_options.inter_op_num_threads)
-                                : nullptr),
-      session_state_(execution_providers_,
-                     session_options.enable_mem_pattern && session_options.execution_mode == ExecutionMode::ORT_SEQUENTIAL,
-                     thread_pool_.get(),
-                     inter_op_thread_pool_.get()),
-      insert_cast_transformer_("CastFloat16Transformer") {
+static Status FinalizeSessionOptions(const SessionOptions& user_provided_session_options,
+                                     const ONNX_NAMESPACE::ModelProto* model_proto,
+                                     /*out*/ SessionOptions& finalalized_session_options) {
+  const logging::Logger& default_logger = logging::LoggingManager::DefaultLogger();
+
+  // By now the environment should have initialized. (It is enforced prior to this.)
+  const Env& env_instance = Env::Default();
+
+  bool read_model_for_ort_config = false;
+
+  // Get the value held by the environment variable - kOrtLoadConfigFromModelEnvVar
+  const std::string& load_config_from_model_env_var_value =
+      env_instance.GetEnvironmentVar(inference_session_utils::kOrtLoadConfigFromModelEnvVar);
+
+  // Ascertain if the model is to be read for the ORT config from the afore parsed env var
+  if (!load_config_from_model_env_var_value.empty()) {
+    // Check if the the env var contains an unsupported value
+    if (load_config_from_model_env_var_value.length() > 1 ||
+        (load_config_from_model_env_var_value[0] != '0' && load_config_from_model_env_var_value[0] != '1')) {
+      std::ostringstream oss;
+      oss << "The only supported values for the environment variable " << inference_session_utils::kOrtLoadConfigFromModelEnvVar
+          << " are '0' and '1'. "
+          << "The environment variable contained the value: " << load_config_from_model_env_var_value;
+      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, oss.str());
+    }
+
+    if (load_config_from_model_env_var_value[0] == '1') {
+      LOGS(default_logger, INFO) << "Reading the provided model for the ORT config";
+      read_model_for_ort_config = true;
+    }
+  }
+
+  // The model is to be read for an ORT config json that may hold some/all session options
+  if (read_model_for_ort_config) {
+    SessionOptions constructed_session_options;
+
+    // In theory we should not hit this condition unless this internal class' APIs are being called incorrectly.
+    // This is a good sanity check to enforce that the model has been parsed prior to looking into it for ort config.
+    if (!model_proto) {
```

Comment:
You can use ORT_ENFORCE to clarify this intent.

Reply:
Sure. Fixed.